### PR TITLE
KAFKA-12462: proceed with task revocation in case of thread in PENDING_SHUTDOWN

### PR DESCRIPTION
**Pull request by ableegoldman**
Should be cherrypicked back to 2.7 & 2.8

### CODE REVIEW

**Comment by ableegoldman** [comment]
@vvcephei @wcarlson5 @lct45

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsRebalanceListener.java [hunk 1]
```diff
@@ -87,7 +87,9 @@ public void onPartitionsRevoked(final Collection<TopicPartition> partitions) {
                   taskManager.activeTaskIds(),
                   taskManager.standbyTaskIds());
 
-        if (streamThread.setState(State.PARTITIONS_REVOKED) != null && !partitions.isEmpty()) {
+        // We need to still invoke handleRevocation if the thread has been told to shut down, but we shouldn't ever
+        // transition away from PENDING_SHUTDOWN once it's been initiated (to anything other than DEAD)
+        if ((streamThread.setState(State.PARTITIONS_REVOKED) != null || streamThread.state() == State.PENDING_SHUTDOWN) && !partitions.isEmpty()) {
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsRebalanceListener.java:92]: do we need to be concerned about the oder these execute?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 2]
```diff
@@ -714,6 +714,13 @@ void runOnce() {
 
         final long pollLatency = pollPhase();
 
+        // Optimization to skip the rest of the processing loop in case the thread was requested to shut down during
+        // the poll phase
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Good idea but I think we do this a few lines down

**Review by wcarlson5 — COMMENTED** [review]
Thanks for the super quick PR. Just had a couple of comments

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 2 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: 🤦‍♀️  Oh wow how did I not see that lol. I'll just bump the log to INFO

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsRebalanceListener.java (hunk 1 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsRebalanceListener.java:92]: I think this is the correct order (assuming you mean the order of `streamThread.setState(State.PARTITIONS_REVOKED) != null` relative to `streamThread.state() == State.PENDING_SHUTDOWN`?) -- if the thread is not in PENDING_SHUTDOWN when it reaches this line, the first condition should return true, which is what we want even if it does get transitioned to PENDING_SHUTDOWN immediately after the transition to PARTITIONS_REVOKED.

**Comment by wcarlson5** [comment]
LGTM

**Comment by ableegoldman** [comment]
One unrelated test failure: `kafka.network.ConnectionQuotasTest.testListenerConnectionRateLimitWhenActualRateAboveLimit()`

Going to merge

**Comment by ableegoldman** [comment]
Merged to trunk and cherrypicked to 2.7 & 2.8 cc @vvcephei

**Comment by wcarlson5** [comment]
Can we cherry pick with back to 2.6 as well?

**Comment by ableegoldman** [comment]
> Can we cherry pick with back to 2.6 as well?

Done

**Pull request merged**